### DISCUSSION TITLE
Como adicionar um json ao Body do dwParams?

### DISCUSSION THREAD

**Discussion by Hawkinks** [OpenSourceCommunityBrasil/REST-DataWare#345] [Q&A] [answered]
Como adicionar um json ao Body do dwParams para fazer um request com o DWClientEvents?
Sei como adicionar os parâmentros na URL para requisição, mas como a rotina que preciso é de Login, não é adequado passar Usuário e Senha pela URL.
Já tentei da seguinte forma, mas deu _Access Violation_:
```
json := TJsonObject.Create;
json.AddPair('usuario', EncodeStrings(EdtUsuario.Text));
json.AddPair('senha', EncodeStrings(EdtSenha.Text));
DM.RESTDWClientEvents.CreateDWParams('login', dwParams);
dwParams.Add(json);
DM.RESTDWClientEvents.SendEvent('login', dwParams, vErrorMessage, vNativeResult)
```
O que estou fazendo errado?
Ou há alguma forma melhor de fazer isso?
Me desculpem por perguntar algo tão básico, mas não encontrei nenhum material sobre isso.
Obrigado!

**Comment by mobius1qwe** [comment]
Qual versão do componente? Delphi ou Lazarus?

**Reply by Hawkinks** [reply]
[v2.0.8](https://github.com/OpenSourceCommunityBrasil/REST-DataWare/wiki/Instala%C3%A7%C3%A3o-no-Delphi)
Delphi.

**Comment by mobius1qwe** [comment, marked as answer]
2.0.7 não é a última, do trunk a última é 2.0.8-29, mas ok.
Pra usar o clientEvents você precisa construir o objeto param antes de adicionar ele no DWParams, tipo assim:

```Delphi
procedure TfPrincipal.cbTableNameEnter(Sender: TObject);
var
  ClientPooler: TRESTDWIdClientPooler;
  Params: TRESTDWParams;
  JSONParam: TJSONParam;
begin
  ConfiguraComponentes;
  
  // Cria os componentes
  ClientPooler := TRESTDWIdClientPooler.Create(nil);
  ClientPooler.AuthenticationOptions := FAuthType;

  Params := TRESTDWParams.Create;
  Params.Encoding := ClientPooler.Encoding;

  // Cria os parâmetros
  JSONParam := TJSONParam.Create(ClientPooler.Encoding);
  JSONParam.ParamName := 'Result';
  JSONParam.ObjectDirection := odOUT;
  JSONParam.ObjectValue := ovString;
  JSONParam.AsString := '';
  Params.Add(JSONParam);

  JSONParam := TJSONParam.Create(ClientPooler.Encoding);
  JSONParam.ParamName := 'Error';
  JSONParam.ObjectDirection := odINOUT;
  JSONParam.ObjectValue := ovBoolean;
  JSONParam.AsBoolean := false;
  Params.Add(JSONParam);

  JSONParam := TJSONParam.Create(ClientPooler.Encoding);
  JSONParam.ParamName := 'Pooler';
  JSONParam.ObjectDirection := odIN;
  JSONParam.ObjectValue := ovString;
  JSONParam.AsString := cbPoolerDB.Selected.Text;
  Params.Add(JSONParam);

  try // Faz a requisição
    ClientPooler.SendEvent('GetTableNames', Params);
    
    // busca o resultado da requisição
    ceTableName.Items.Delimiter := '|';
    ceTableName.Items.DelimitedText := Params.ItemsString['Result'].AsString;
  finally
    Params.Free;
    ClientPooler.Free;
  end;
end;
```
Esse código acima aí tá no demo SimpleClient

**Comment by Hawkinks** [comment]
Deu certo. Obrigado!
Agora no meu servidor eu pensava estar correto.
Para acessar o conteúdo Body do Params por lá eu estava fazendo assim:
```
body:= TJSONObject.ParseJSONValue(TEncoding.UTF8.GetBytes(Params.RawBody.AsString),0) as TJSONValue;
usuario.Usuario := body.GetValue<String>('usuario');
usuario.Senha := body.GetValue<String>('senha');
```
Testei um request no PostMan colocando os dados no rawbody `{"usuario":"fulano","senha":"1234"}` e funciona normalmente.

Parece que ainda estou errando em algo básico.

**Reply by Hawkinks** [reply]
Esqueci de mencionar que dessa forma quando a aplicação cliente envia o request dá o erro _Invalid RESTDWParam_ no servidor.

**Reply by Hawkinks** [reply]
Encontrei uma forma, mas me pareceu uma "gato". rsrsrs
O correto seria isso mesmo?
```
    body:= TJSONObject.ParseJSONValue(Params.Items[8].ToJSON) as TJSONValue;
    usuario.Usuario := DecodeStrings(body.GetValue<String>('USUARIO'));
    body:= TJSONObject.ParseJSONValue(Params.Items[9].ToJSON) as TJSONValue;
    usuario.Senha := DecodeStrings(body.GetValue<String>('SENHA'));
```

**Comment by mobius1qwe** [comment]
Para poder obter o body da requisição, você precisa enviar de outra forma, como você tá usando ClientEvents, ele não faz distinção de header e body, joga tudo no mesmo lugar, você precisaria filtrar pelo nome do param de acordo com a minha segunda resposta com o código.

Vamos considerar que você esteja fazendo algo parecido com isso aqui:
```Delphi
  JSONParam := TJSONParam.Create(ClientPooler.Encoding);
  JSONParam.ParamName := 'meusdados';
  JSONParam.ObjectDirection := odIN;
  JSONParam.ObjectValue := ovString;
  JSONParam.AsString := meuJSON.ToJSON;
  Params.Add(JSONParam);
```

Daí no servidor, você vai puxar essa informação através do `Params.ItemsString['meusdados'].AsString;` pois com ClientEvents não existe distinção de header e body.

**Comment by Hawkinks** [comment]
Bacana!
Inicialmente eu havia feito dessa forma, mas eu pensei que estava fazendo algo sem segurança. Esclareceu muito essa informação que ele não distingue header e body.

Só mais uma coisa. O que tenho que configurar para descriptografar os dados do Params automaticamente quando recebido no servidor?
O cliente já está enviando criptografado, só falta o servidor descriptografar.
No meu ClientPooler a key do CriptOptions está setada como 'RDWBASEKEY256' e ativo, exatamente da mesma forma que está no ServicePooler. O que está faltando?

**Comment by mobius1qwe** [comment]
se você tiver fazendo dessa forma aqui:
` body:= TJSONObject.ParseJSONValue(Params.Items[8].ToJSON) as TJSONValue;`
não vai funcionar, porque a função que decodifica o conteúdo do Params é a .AsString, daí você precisaria fazer
` body:= TJSONObject.ParseJSONValue(Params.Items[8].AsString) as TJSONValue;`
pra conseguir decodificar corretamente o valor.

**Reply by Hawkinks** [reply]
Tanto assim:
`body:= TJSONObject.ParseJSONValue(Params.Items[8].AsString) as TJSONValue;`
Como assim:
`Params.ItemsString['meusdados'].AsString;`
Não está decodificando. 
Só mostra caracteres "normais" quando desmarco a opção Use em CriptOptions no ClientPooler.

**Comment by mobius1qwe** [comment]
Bom, internamente ele faz essa validação e decripta se tiver configurado pra decriptar, tenta testar isso na 2.1 e veja se o problema persiste, se persistir, recomendo abrir uma tarefa detalhando o problema.

**Comment by Hawkinks** [comment]
Testei na versão 2.1 e o problema persiste.
Vou abrir uma tarefa sobre o assunto.

Muito obrigado pela ajuda!

**Comment by Hawkinks** [comment]
Estou prestes a abrir a tarefa, mas ainda temo que eu ainda seja o problema.
Vou mostrar aqui como estou fazendo na tarefa, e se possível, você me diz se está correto e se realmente deveria ter um output diferente, pode ser?
Daí você me confirmando, eu fecho essa discussão e abro a tarefa em seguida.

No ClientPooler em CriptOptions defina como:
```
Key: rdwAES256
Use: true
```
Envie uma requisição com ClientEvents:
```
json := TJsonObject.Create;
json.AddPair('USUARIO', 'fulano');
json.AddPair('SENHA', '1234');
DM.RESTDWClientEvents.CreateDWParams('login', dwParams);
dwParams.ItemsString['dadoslogin'].AsString := json.ToJSON;
DM.RESTDWClientEvents.SendEvent('login', dwParams, vErrorMessage, vNativeResult)
```
Em CriptOptions no ServicePooler defina da mesma forma que ClientPooler.
Receba no evento 'login' em ServerEvents:
```
body := TJSONObject.ParseJSONValue(Params.ItemsString['dadoslogin'].AsString) as TJSONValue
usuario := body.GetValue<String>('USUARIO');
senha := body.GetValue<String>('SENHA');
```
Irá obter o erro: _Value 'USUARIO' not found_.
Para testar visualmente:
`showmessage(Params.ItemsString['dadoslogin'].AsString);`
Verá os dados ainda codificados.

**Comment by mobius1qwe** [comment]
```
json := TJsonObject.Create;
json.AddPair('USUARIO', 'fulano');
json.AddPair('SENHA', '1234');
DM.RESTDWClientEvents.CreateDWParams('login', dwParams); <=== aqui era pra ser 'dadoslogin'
dwParams.ItemsString['dadoslogin'].AsString := json.ToJSON;  <==== igual aqui
DM.RESTDWClientEvents.SendEvent('login', dwParams, vErrorMessage, vNativeResult)
```
Mas porque você não faz da forma que mostrei acima? Dessa forma aí parece confuso...
https://github.com/OpenSourceCommunityBrasil/REST-DataWare/discussions/345#discussioncomment-6020161

**Reply by Hawkinks** [reply]
É porque tentei colocar o mais resumido possível.
Como assim, em CreateDWParams tem que colocar o nome do parâmetro?
Nele eu não tenho que informar o nome do evento para ele criar o dwParams de acordo com os parâmetros já criados nesse evento? Estou enganado?

Mas você tem razão, vou colocar do outro jeito, talvez seja melhor para o entendimento.
```
dwParams := TRESTDWParams.Create;
JSONParam := TJSONParam.Create(dm.RESTDWIdClientPooler.Encoding);
JSONParam.ParamName := 'dadoslogin';
JSONParam.ObjectDirection := odIn;
JSONParam.ObjectValue := ovString;
JSONParam.AsString := json.tojson;
dwParams.Add(JSONParam);
```
